docs/madr/decisions/040-meshservice-hostname-vips.md

```diff
@@ -0,0 +1,128 @@
+Kuma VIP is a non-routable IP used only by Envoy to indicate destination (like a host header, but on L4). We currently do this with 240.0.0.0/4 CIDR.
+
+Kuma VIPs are assigned asynchronously with MeshService creation. Assigning Kuma VIPs as sync operation on MeshService creation would create a race condition on VIP allocation.
+
+VIP assigning process (single-threaded):
+* List MeshServices
+* If a MeshService has no VIP, assign a Kuma VIP. Kubernetes Services would have a VIP already (ClusterIP), so we don’t need to assign an extra VIP.
+* To assign a VIP, we build an IPAM state reserving every address we already assigned and allocate the next address.
+  To solve the problem of “assigning immediately” a VIP of an old service to a new Service we can also store in memory recently removed VIPs for X seconds and set TTL of DNS responses for X seconds.
+  The current value of 30s is too long, especially that the DNS server is run locally in kuma-dp. Leader election should take more than X seconds, so if we were to lose this list, it’s not a problem.
+
+Following this process, we don’t have to store VIPs in any central place like we do currently (VIP ConfigMap for each Mesh).
+We also assign VIPs for ExternalServices, but this can work as long as we consider both ExternalServices and MeshServices for this process.
+
+To avoid collision with an existing VIPs mechanism, we can take a completely new CIDR (241.0.0.0/4).
+
+It's not expected for service to have multiple VIPs, but we want expose this possibility for the future (therefore `vips` field is a list).
+
+### Hostname Generator
+
+MeshService has to be addressable with a hostname.
+In case of services in a Kubernetes cluster, we don't have to do anything because we can just rely on Kube DNS that resolves to a Kube VIP we placed for MeshService.
+
+In case of services in a Universal cluster, currently we provide two ways:
+* `<kuma.io/service>.mesh` address that resolves to Kuma VIP
+* VirtualOutbound resource to provide custom addresses
+
+Hostnames should be managed only by MeshOperator. Otherwise, we get into security issues such as one service owner can override a hostname of other service.
+Hostnames are created by convention in the majority of cases - Kubernetes hostnames are also such example.
+It is still possible for service owner to create unique hostname for their app by asking reaching out to mesh operator.
+
+To manage hostname, we introduce new object called `HostnameGenerator`
+
+```yaml
+type: HostnameGenerator
+name: k8s-zone-hostnames
+spec:
+  meshServiceSelector:
+    matchLabels:
+      kuma.io/zone: east 
+  template: {{ label "k8s.kuma.io/service-name" }}.{{ label "k8s.kuma.io/namespace" }}.svc.cluster.{{ label "kuma.io/zone" }}
+```
+
+`HostnameGenerator` is a namespaced-scoped object on Kubernetes, but can only be applied in system namespace.
```

Comment:
Why is it namespaced?

Reply:
It's just safer. We can limit the namespace to a system namespace, but I'd rather have it just in case.

Reply:
added to madr